In [20]:
%matplotlib inline
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from matplotlib import pyplot as plt
from sklearn.datasets import make_regression
from sklearn.preprocessing import StandardScaler

In [21]:
file_path = Path("Resources/lending_data.csv")
df = pd.read_csv(file_path)
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

## Prediction
### My prediction is that Logistic Regression will perform better than Random Tree Classifier. The reason why I believe that is because Logistic Regression tends to perform better on categorical data, which in this case our data is categorical and descreet, with loan status being category 0 and 1. Logistic Regression will not predict the exact category the observation should be in, but gives a probability of each observation falls into the category ‘1’.


In [22]:
# Assign the data to X and y
# Note: Sklearn requires a two-dimensional array of values
# so we use reshape() to create this

X = df[['loan_size','interest_rate','borrower_income', 'debt_to_income', 'num_of_accounts', 'derogatory_marks', 'total_debt']]
y = df['loan_status']

print("Shape: ", X.shape, y.shape)

Shape:  (77536, 7) (77536,)


In [4]:
# Split the data into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [5]:
X_test.head()

loan_size  interest_rate  borrower_income  debt_to_income  \
60914    12600.0          8.469            60300        0.502488   
36843     9800.0          7.289            49200        0.390244   
1966     10900.0          7.770            53700        0.441341   
70137    10700.0          7.666            52700        0.430740   
27237     9900.0          7.353            49800        0.397590   

       num_of_accounts  derogatory_marks  total_debt  
60914                6                 1       30300  
36843                4                 0       19200  
1966                 5                 1       23700  
70137                5                 1       22700  
27237                4                 0       19800

In [6]:
# Train a Logistic Regression model print the model score
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier

LogisticRegression()

In [7]:
classifier.fit(X_train, y_train)

LogisticRegression()

In [8]:
# Print the accuracy score for the test data

print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.9921240885954051
Testing Data Score: 0.9918489475856377


In [9]:
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
classifier2 = LogisticRegression()
classifier2.fit(X_train_scaled, y_train)
print(f"Training Data Score: {classifier2.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {classifier2.score(X_test_scaled, y_test)}")

Training Data Score: 0.9942908240473243
Testing Data Score: 0.9936545604622369


In [10]:
# Train a Random Forest Classifier model and print the model score

from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train, y_train)
print(f'Training Score: {clf.score(X_train, y_train)}')
print(f'Testing Score: {clf.score(X_test, y_test)}')

Training Score: 0.9975409272252029
Testing Score: 0.991642591828312


In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [12]:
clf2 = RandomForestClassifier(random_state=1, n_estimators=500).fit(X_train_scaled, y_train)
print(f'Training Score: {clf2.score(X_train_scaled, y_train)}')
print(f'Testing Score: {clf2.score(X_test_scaled, y_test)}')

Training Score: 0.9975409272252029
Testing Score: 0.9917457697069748


## Results: 
### Both models were very close in their perfomance, but the Logistic Regression model performed slightly better on both, unscaled and scaled original data. 
### When it comes to unscaled data, the Logistic Regression testing data score is 0.9918 and Random Forest Classifier testing score is 0.9916. 
### When it comes to scaled testing scores, Logistic Regression testing data score is 0.9937 and Random Forest Classifier testing score is 0.9917. 
### These results concur with my prediction, but with this much accuracy from both models, either one can be used to predict whether a loan will be approved or not.